### PR TITLE
[FLINK-15947] Finish moving scala expression DSL to flink-table-api-scala

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change
As per discussion here: http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-TABLE-Issue-with-package-structure-in-the-Table-API-td37623.html

This PR:
* moves the scala expression DSL to org.apache.flink.table.api package in
table-api-scala module
* moves all classes from org.apache.flink.table.api.scala and
org.apache.flink.table.api.java packages to
org.apache.flink.table.api.bridge.scala and
org.apache.flink.table.api.bridge.java accordingly and drops the former
packages

**This is a breaking API change!!**

## Verifying this change

All the existing tests pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
This change would break nearly all users setup. Nevertheless I think it is a very important change that significantly improves usability of the Table API.

Let me know what do you think about performing that change for 1.11. @aljoscha @twalthr @KurtYoung @wuchong

**Comment by twalthr** [comment]
Thanks for opening a PR for this topic @dawidwys. I wish we could avoid this but I don't see a possibility to prevent this change. The clash with the Scala macro and the annoying `_root_.scala` are very annoying API design fails that almost every user will observe eventually. Looking at the `WordCount` example shows easy and understandable imports now:

```
import org.apache.flink.table.api._
import org.apache.flink.table.api.bridge.scala._
```

And this is what we should aim for. Given that the user base is growing rapidly, we do this rather sooner than later. Given the amount of API that we have already changed, a simple update of the import line on the top is a rather easy change to apply.

**Comment by dawidwys** [comment]
@flinkbot run azure

**Comment by aljoscha** [comment]
This is a tough tradeoff but I think we need to break existing setups now in order to fix the problems that this causes. 👌

**Comment by wuchong** [comment]
I think if we decide to change the package path, 1.11 is a good chance to do it, because we also change the execute() behavior in this version which forces users to update their programs. Otherwise, it might be impossible to do it in the future. The question is whether we should change it or not. This is really a tough decision (will break almost all the users program) :pensive:, is it worth to do it?

**Comment by dawidwys** [comment]
@wuchong My take is that it is inevitable at some point. With the current structure there are three issues:
1. users of the unified `TableEnvironment` have to depend on a bridging module for the expressionDsl
2. if users import `org.apache.flink.table.api._` they have to prefix all classes from `java` & `scala` with `_root_`
3. moreover in point 2. it breaks `TypeInformation` extraction for all such classes.

Therefore the question is rather if we do it now or in a next release...

**Comment by dawidwys** [comment]
To give the discussion some time limit, if there are no objections (meaning people clearly against it) till tomorrow 10 am CEST I will introduce that change (if the release managers agree @pnowojski  @zhijiangW).

**Comment by wuchong** [comment]
Thanks @dawidwys for the explanation. As I said above, if we decide to change it, and we have to do it, then we should do it in this release. So I'm fine with this.

File: flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/package.scala [hunk 1]
```diff
@@ -39,7 +37,7 @@ import org.apache.flink.table.api.{ImplicitExpressionConversions, ImplicitExpres
   *
   * Please refer to the website documentation about how to construct and run table programs.
   */
-package object api /* extends ImplicitExpressionConversions */ {
+package object api extends ImplicitExpressionConversions {
 
   // This package object should extend from ImplicitExpressionConversions but would clash with
```
Comment (twalthr) [inline flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/package.scala:42]: update comment

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/expressions/package.scala [hunk 2]
```diff
@@ -21,8 +21,8 @@ package org.apache.flink.table
  * This package contains the base class of AST nodes and all the expression language AST classes.
  * Expression trees should not be manually constructed by users. They are implicitly constructed
  * from the implicit DSL conversions in
- * [[org.apache.flink.table.api.scala.ImplicitExpressionConversions]] and
- * [[org.apache.flink.table.api.scala.ImplicitExpressionOperations]]. For the Java API,
+ * [[org.apache.flink.table.api.bridge.scala.ImplicitExpressionConversions]] and
+ * [[org.apache.flink.table.api.bridge.scala.ImplicitExpressionOperations]]. For the Java API,
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/expressions/package.scala:25]: update last sentence

File: flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkILoop.scala [hunk 3]
```diff
@@ -23,7 +23,7 @@ import org.apache.flink.api.java.{JarHelper, ScalaShellEnvironment, ScalaShellSt
 import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
 import org.apache.flink.configuration.Configuration
 import org.apache.flink.table.api.EnvironmentSettings
-import org.apache.flink.table.api.scala.{BatchTableEnvironment, StreamTableEnvironment}
+import org.apache.flink.table.api.bridge.scala.{BatchTableEnvironment, StreamTableEnvironment}
```
Comment (twalthr) [inline flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkILoop.scala:26]: Update the default imports in the lower part of this class.

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/package.scala [hunk 4]
```diff
@@ -50,7 +50,7 @@ import _root_.scala.language.implicitConversions
   * Please refer to the website documentation about how to construct and run table programs that are
   * connected to the DataStream API.
   */
-package object scala extends ImplicitExpressionConversions {
+package object scala {
 
   // This package object should not extend from ImplicitExpressionConversions but would clash with
```
Comment (twalthr) [inline flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/package.scala:55]: update comment

File: flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala [hunk 5]
```diff
@@ -33,7 +33,7 @@ import java.sql.{Date, Time, Timestamp}
 import java.time.{LocalDate, LocalDateTime, LocalTime}
 import java.util.{List => JList, Map => JMap}
```
Comment (twalthr) [inline flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/expressionDsl.scala:34]: nit: Shall we move `ImplicitExpressionConversions` and `ImplicitExpressionOperations` into dedicated files? Personally, I don't like `expressionDsl.scala` much. `TableConversions` and `DataStreamConvertions` are also dedicated files.

File: docs/dev/table/common.md [hunk 6]
```diff
@@ -552,7 +552,9 @@ val revenue = orders
 // execute query
 {% endhighlight %}
 
-**Note:** The Scala Table API uses Scala Symbols, which start with a single tick (`'`) to reference the attributes of a `Table`. The Table API uses Scala implicits. Make sure to import `org.apache.flink.api.scala._` and `org.apache.flink.table.api.scala._` in order to use Scala implicit conversions.
+**Note:** The Scala Table API uses Scala Symbols, which start with a single tick (`'`) to reference the attributes of a `Table`. The Table API uses Scala implicits. Make sure to import
```
Comment (twalthr) [inline docs/dev/table/common.md:555]: `The Scala Table API uses Scala Symbols, which start with a single tick (`'`)`, update?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/expressions/package.scala (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/expressions/package.scala:25]: Or shall I just remove this file? I don't see any reason for it.

**Comment by dawidwys** [comment]
Thank you all for joining the discussion!
Thanks @twalthr for the review. I'll run the tests and after that I will merge that change.

**Pull request closed without merging**